### PR TITLE
[GEP-24] Introduce `ShootManagedIssuer` feature gate

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area security usability
/kind enhancement

**What this PR does / why we need it**:
This PR introduces a feature gate called `ShootManagedIssuer` that is meant to guard the functionality related to https://github.com/gardener/gardener/issues/9157 until it is fully implemented. Right now the most crucial part - the `metadata/discovery` server is still not implemented. In addition to that if one annotates a shoot cluster, but the shoot service account issuer hostname is not configured in the Garden cluster the reconciliation will fail and operator intervention will be required. This is why I propose that we guard these features until they become fully available. I plan to promote the feature gate to beta once the `gardener-discovery-server` is implemented and integrated with the `gardener-operator`. After a few more releases we can promote this feature gate to GA and remove it.
 
 
 cc @petersutter 
 
 /cc @vpnachev @rfranzke @timuthy 
 
**Which issue(s) this PR fixes**: 
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature operator
A new gardenlet feature gate called `ShootManagedIssuer` was introduced. This feature gate guards the functionality described in [GEP-24](https://github.com/gardener/gardener/blob/master/docs/proposals/24-shoot-oidc-issuer.md) until all of the components mentioned in the enhancement proposal are implemented by Gardener.
```

### CODE REVIEW

**Review by vpnachev — COMMENTED** [review]
I just have two questions:
- should the feature gate be documented in the managed issuer section in https://github.com/gardener/gardener/blob/master/docs/operations/configuration.md#system-configuration ?
- should the feature gate be enabled in the example config ?https://github.com/gardener/gardener/blob/4be4ddb95822713c78d012ebc9ab2dfea325904d/example/20-componentconfig-gardenlet.yaml#L111-L119

**Comment by dimityrmirchev** [comment]
> I just have two questions:
> 
> * should the feature gate be documented in the managed issuer section in https://github.com/gardener/gardener/blob/master/docs/operations/configuration.md#system-configuration ?
> * should the feature gate be enabled in the example config ?https://github.com/gardener/gardener/blob/4be4ddb95822713c78d012ebc9ab2dfea325904d/example/20-componentconfig-gardenlet.yaml#L111-L119

Thanks @vpnachev 
I mentioned the feature gate in both places. I think we can leave it to false in the gardenlet config since it is an alpha feature.

**Review by vpnachev — APPROVED** [review]
/lgtm

File: docs/operations/configuration.md [hunk 1]
```diff
@@ -75,7 +75,7 @@ When the `gardenlet` starts, it scans the `garden` namespace of the garden clust
 * **Global monitoring secrets** (optional) - contains basic authentication credentials for the Prometheus aggregating metrics for all clusters.
   * These secrets are synced to each seed cluster and used to gain access to the aggregate monitoring components.
 
-* **Shoot Service Account Issuer secret** (optional) - contains the configuration needed to centrally configure gardenlets in order to implement [GEP-24](../proposals/24-shoot-oidc-issuer.md). Please see [the example configuration](../../example/10-secret-shoot-service-account-issuer.yaml) for more details.
+* **Shoot Service Account Issuer secret** (optional) - contains the configuration needed to centrally configure gardenlets in order to implement [GEP-24](../proposals/24-shoot-oidc-issuer.md). Please see [the example configuration](../../example/10-secret-shoot-service-account-issuer.yaml) for more details. In addition to that the [`ShootManagedIssuer`](../deployment/feature_gates.md#list-of-feature-gates) gardenlet feature gate should be enabled in order for configurations to take effect.
```
Comment (rfranzke) [inline docs/operations/configuration.md]:
```suggestion
* **Shoot Service Account Issuer secret** (optional) - contains the configuration needed to centrally configure gardenlets in order to implement [GEP-24](../proposals/24-shoot-oidc-issuer.md). Please see [the example configuration](../../example/10-secret-shoot-service-account-issuer.yaml) for more details. In addition to that, the [`ShootManagedIssuer`](../deployment/feature_gates.md#list-of-feature-gates) gardenlet feature gate should be enabled in order for configurations to take effect.
```

File: extensions/.import-restrictions [hunk 2]
```diff
@@ -23,4 +23,6 @@ rules:
   - github.com/gardener/gardener/pkg/logger
   - github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector/references
   - github.com/gardener/gardener/pkg/utils
-  - github.com/gardener/gardener/third_party/mock
\ No newline at end of file
+  - github.com/gardener/gardener/third_party/mock
+  # TODO(dimityrmirchev): Remove the below rule when the ShootManagedIssuer feature gate is promoted to GA.
+  - github.com/gardener/gardener/pkg/features
```
Comment (rfranzke) [inline extensions/.import-restrictions]: Why is this needed here? 😕

File: pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go [hunk 3]
```diff
@@ -428,9 +429,11 @@ func (r *Reconciler) runReconcileShootFlow(ctx context.Context, o *operation.Ope
...
 		})
 		_ = g.Add(flow.Task{
-			Name:         "Sync public service account signing keys to Garden cluster",
-			Fn:           botanist.SyncPublicServiceAccountKeys,
-			SkipIf:       o.Shoot.HibernationEnabled || !v1beta1helper.HasManagedIssuer(botanist.Shoot.GetInfo()),
+			Name: "Sync public service account signing keys to Garden cluster",
+			Fn:   botanist.SyncPublicServiceAccountKeys,
+			SkipIf: o.Shoot.HibernationEnabled ||
+				!features.DefaultFeatureGate.Enabled(features.ShootManagedIssuer) ||
+				!v1beta1helper.HasManagedIssuer(botanist.Shoot.GetInfo()),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go:436]: Should the synced keys be deleted in case somebody disables the feature gate after it was enabled?

File: pkg/gardenlet/operation/botanist/advertisedaddresses_test.go [hunk 4]
```diff
@@ -183,6 +186,47 @@ var _ = Describe("AdvertisedAddresses", func() {
...
+				}, {
+					Name: "internal",
+					URL:  "https://api.baz.foo",
+				},
+				{
+					Name: "service-account-issuer",
+					URL:  "https://api.baz.foo",
+				},
+			}))
+		})
```
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/advertisedaddresses_test.go:228]: I see no changes in the `ToAdvertisedAddresses` function, can you explain how this unit test is relevant/testing anything? 👀

Comment (dimityrmirchev) [reply pkg/gardenlet/operation/botanist/advertisedaddresses_test.go:228]: The function `ToAdvertisedAddresses` always reports the `service-account-issuer` address if possible, not only if the managed issuer is enabled. Then it calls `computeKubeAPIServerServiceAccountConfig` which uses the feature gate to determine if the annotation should be taken into consideration. The feature gate is set to `false` in the test and this is the reason why the `service-account-issuer` address is equal to the default `internal` address.

File: extensions/.import-restrictions (hunk 2 above)
Comment (dimityrmirchev) [reply extensions/.import-restrictions]: Seems to not be needed after the latest rebase https://github.com/gardener/gardener/pull/9489/commits/e42876ef942588dffab89201ccc06205d605b84b

File: pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go (hunk 3 above)
Comment (dimityrmirchev) [reply pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go:436]: Good point. Done with https://github.com/gardener/gardener/pull/9489/commits/6b28c42bb90a4938b825414618de13245b539b26
